### PR TITLE
Add SSL_{CLIENT,SERVER}_B64CERT

### PR DESCRIPTION
**Pull request by notroj**
Add SSL_{CLIENT,SERVER}_B64CERT to export the base64url(DER) of the server/client cert.

PR: 65169

### CODE REVIEW

File: modules/ssl/ssl_engine_vars.c [hunk 1]
```diff
@@ -71,6 +76,21 @@ static int ssl_is_https(conn_rec *c)
     return sslconn && sslconn->ssl;
 }
 
+#if APR_VERSION_AT_LEAST(1,7,0)
```
Comment (michael-o) [inline modules/ssl/ssl_engine_vars.c]: Is this really available in 1.7.0+ only?

File: modules/ssl/ssl_engine_vars.c [hunk 2]
```diff
@@ -71,6 +76,21 @@ static int ssl_is_https(conn_rec *c)
...
+static const char *ssl_var_lookup_ssl_cert_b64url(apr_pool_t *p, X509 *xs)
```
Comment (michael-o) [inline modules/ssl/ssl_engine_vars.c]: The function name is wrong now.

**Review by michael-o — CHANGES_REQUESTED** [review]

File: modules/ssl/ssl_engine_vars.c (hunk 2 above)
Comment (notroj) [reply modules/ssl/ssl_engine_vars.c]: Thanks a lot for the review!

File: modules/ssl/ssl_engine_vars.c (hunk 1 above)
Comment (notroj) [reply modules/ssl/ssl_engine_vars.c]: This API was.   But there are three to choose from, I switched to using the OpenSSL API instead int 531195b - thanks for reviewing!

File: modules/ssl/ssl_engine_vars.c [hunk 3]
```diff
@@ -593,8 +623,13 @@ static const char *ssl_var_lookup_ssl_cert(apr_pool_t *p, request_rec *r, X509 *
         result = (nid == NID_undef) ? "UNKNOWN" : OBJ_nid2ln(nid);
     }
     else if (strcEQ(var, "CERT")) {
-        result = ssl_var_lookup_ssl_cert_PEM(p, xs);
+        result = ssl_var_lookup_ssl_cert_data(p, xs, 1);
     }
+#if APR_VERSION_AT_LEAST(1,7,0)
```
Comment (michael-o) [inline modules/ssl/ssl_engine_vars.c:632]: Why does this still require APR 1.7.0+?

File: modules/ssl/ssl_engine_kernel.c [hunk 4]
```diff
@@ -1607,6 +1607,15 @@ int ssl_hook_Fixup(request_rec *r)
...
     }
 
+    if (dc->nOptions & SSL_OPT_EXPORTCB64DATA) {
+        val = ssl_var_lookup(r->pool, r->server, r->connection,
+                             r, "SSL_SERVER_B64CERT");
+        apr_table_setn(env, "SSL_SERVER_B64CERT", val);
+
+        val = ssl_var_lookup(r->pool, r->server, r->connection,
+                             r, "SSL_CLIENT_B64CERT");
+        apr_table_setn(env, "SSL_CLIENT_B64CERT", val);
```
Comment (michael-o) [inline modules/ssl/ssl_engine_kernel.c]: I still miss the entire chain for the sake of completeness.

File: modules/ssl/ssl_private.h [hunk 5]
```diff
@@ -356,6 +356,7 @@ APLOG_USE_MODULE(ssl);
 #define SSL_OPT_STRICTREQUIRE  (1<<5)
 #define SSL_OPT_OPTRENEGOTIATE (1<<6)
 #define SSL_OPT_LEGACYDNFORMAT (1<<7)
+#define SSL_OPT_EXPORTCB64DATA (1<<8)
```
Comment (michael-o) [inline modules/ssl/ssl_private.h:359]: Shouldn't this rather read `SSL_OPT_EXPORTBASE64CERTDATA`?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: modules/ssl/ssl_engine_vars.c [hunk 6]
```diff
@@ -477,7 +477,7 @@ static const char *ssl_var_lookup_ssl(apr_pool_t *p, const SSLConnRec *sslconn,
         sk = SSL_get_peer_cert_chain(ssl);
         result = ssl_var_lookup_ssl_cert_chain(p, sk, var+18, 1);
     }
-    else if (ssl != NULL && strlen(var) > 18 && strcEQn(var, "CLIENT_B64CERT_CHAIN_", 18)) {
+    else if (ssl != NULL && strlen(var) > 21 && strcEQn(var, "CLIENT_B64CERT_CHAIN_", 21)) {
         sk = SSL_get_peer_cert_chain(ssl);
         result = ssl_var_lookup_ssl_cert_chain(p, sk, var+18, 0);
```
Comment (michael-o) [inline modules/ssl/ssl_engine_vars.c]: 21 here too?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: modules/ssl/ssl_engine_vars.c (hunk 6 above)
Comment (notroj) [reply modules/ssl/ssl_engine_vars.c]: That would explain why it wasn't working properly.  Thank you @michael-o

**Pull request closed without merging**